FreeBSD9/contrib/opie/opiekey.1
.\" opiekey.1: Manual page for the opiekey(1) program.
.\"
.\" %%% portions-copyright-cmetz-96
.\" Portions of this software are Copyright 1996-1999 by Craig Metz, All Rights
.\" Reserved. The Inner Net License Version 2 applies to these portions of
.\" the software.
.\" You should have received a copy of the license with this software. If
.\" you didn't get a copy, you may request one from <license@inner.net>.
.\"
.\" Portions of this software are Copyright 1995 by Randall Atkinson and Dan
.\" McDonald, All Rights Reserved. All Rights under this copyright are assigned
.\" to the U.S. Naval Research Laboratory (NRL). The NRL Copyright Notice and
.\" License Agreement applies to this software.
.\"
.\"	History:
.\"
.\"	Modified by cmetz for OPIE 2.3. Added -t documentation. Removed
.\"	   opie-bugs pointer. Removed opie-md5 and opie-md4 names. Fixed
.\"	   a bolding bug. Added -f flag. Added escapes on flags. Minor
.\"        editorial changes. Updated example.
.\"     Modified by cmetz for OPIE 2.2. Removed MJR DES documentation.
.\"        Re-worded retype documentation. Added opiegen reference.
.\"        Added -x documentation.
.\"	Modified at NRL for OPIE 2.0.
.\"	Written at Bellcore for the S/Key Version 1 software distribution
.\"		(key.1).
.\"
.\"	$FreeBSD: release/9.0.0/contrib/opie/opiekey.1 187920 2009-01-30 15:43:55Z gabor $
.ll 6i
.pl 10.5i
.lt 6.0i
.TH OPIEKEY 1 "February 20, 1996"
.AT 3
.SH NAME
opiekey, otp-md4, otp-md5 \- Programs for computing responses to OTP challenges.

.SH SYNOPSIS
.B opiekey
|
.B otp-md4
|
.B otp-md5
[\-v] [\-h] [\-f] [\-x]
.sp 0
[\-t
.I
type
] [\-4|\-5]
[\-a] [\-n 
.I count
] 
.I sequence_number seed
.sp 0

.SH DESCRIPTION
.I opiekey
takes the optional count of the number of responses to
print along with a (maximum) sequence number and seed as command line
args. It prompts for the user's secret pass phrase and produces an OPIE
response as six words. If compiled to do so, it can prompt for the user's
secret pass phrase twice to help reduce errors due to mistypes. The second
password entry can be circumvented by entering only an end of line.
.I opiekey
is downward compatible with the 
.IR key (1)
program from the Bellcore S/Key Version 1 distribution and several of its
variants.

.SH OPTIONS
.TP
.B \-v
Display the version number and compile-time options, then exit.
.TP
.B \-h
Display a brief help message and exit.
.TP
.B \-4, \-5
Selects MD4 or MD5, respectively, as the response generation algorithm. The
default for otp-md4 is MD4 and the default for opie-md5 is MD5. The default
for opiekey depends on compile-time configuration, but should be MD5. MD4 is
compatible with the Bellcore S/Key Version 1 distribution.
.TP
.B \-f
Force
.I opiekey
to continue, even where it normally shouldn't. This is currently used to
force opiekey to operate in even from terminals it believes to be insecure.
It can also allow users to disclose their secret pass phrases to attackers.
Use of the -f flag may be disabled by compile-time option in your particular
build of OPIE.
.TP
.B \-a
Allows you to input an arbitrary secret pass phrase, instead of running checks
against it. Arbitrary currently does not include '\\0' or '\\n' characters. This
can be used for backwards compatibility with key generators that do not check
passwords.
.TP
.B \-n <count>
the number of one time access passwords to print.
The default is one.
.TP
.B \-x
Output the OTPs as hexadecimal numbers instead of six words.
.TP
.B \-t <type>
Generate an extended response of the specified type. Supported types are:
.sp 1
word			six-word
.sp 0
hex			hexadecimal
.sp 0
init			hexadecimal re-initialization
.sp 0
init-word		six-word re-initialization
.sp 1
The re-initialization responses
.I always
generate the simple active attack protection.
.TP
.SH EXAMPLE
.sp 0
wintermute$ opiekey \-5 \-n 5 495 wi01309
.sp 0
Using MD5 algorithm to compute response.
.sp 0
Reminder: Don't use opiekey from telnet or dial-in sessions.
.sp 0
Enter secret pass phrase:
.sp 0
491: HOST VET FOWL SEEK IOWA YAP  
.sp 0
492: JOB ARTS WERE FEAT TILE IBIS 
.sp 0
493: TRUE BRED JOEL USER HALT EBEN
.sp 0
494: HOOD WED MOLT PAN FED RUBY   
.sp 0
495: SUB YAW BILE GLEE OWE NOR    
.sp 0
wintermute$
.LP

.SH BUGS
.BR opiekey(1)
can lull a user into revealing his/her password when remotely logged in, thus
defeating the purpose of OPIE. This is especially a problem with xterm. 
.BR opiekey(1)
implements simple checks to reduce the risk of a user making
this mistake. Better checks are needed.
.LP

.SH SEE ALSO
.BR ftpd (8),
.BR login (1),
.BR opie (4),
.BR opiepasswd (1),
.BR opieinfo (1),
.BR opiekeys (5),
.BR opieaccess (5),
.BR su (1)

.SH AUTHOR
Bellcore's S/Key was written by Phil Karn, Neil M. Haller, and John S. Walden
of Bellcore. OPIE was created at NRL by Randall Atkinson, Dan McDonald, and
Craig Metz.

S/Key is a trademark of Bell Communications Research (Bellcore).

.SH CONTACT
OPIE is discussed on the Bellcore "S/Key Users" mailing list. To join,
send an email request to:
.sp
skey-users-request@thumper.bellcore.com


FreeBSD9/contrib/opie/opiekey.c
/* opiekey.c: Stand-alone program for computing responses to OTP challenges.

 Takes a sequence number and seed (presumably from an OPIE challenge)
 as command line arguments, prompts for the user's secret pass phrase,
 and outputs a response.

%%% portions-copyright-cmetz-96
Portions of this software are Copyright 1996-1999 by Craig Metz, All Rights
Reserved. The Inner Net License Version 2 applies to these portions of
the software.
You should have received a copy of the license with this software. If
you didn't get a copy, you may request one from <license@inner.net>.

Portions of this software are Copyright 1995 by Randall Atkinson and Dan
McDonald, All Rights Reserved. All Rights under this copyright are assigned
to the U.S. Naval Research Laboratory (NRL). The NRL Copyright Notice and
License Agreement applies to this software.

	History:

	Modified by cmetz for OPIE 2.4. Use struct opie_key for key blocks.
	Modified by cmetz for OPIE 2.31. Renamed "init" and RESPONSE_INIT
	        to "init-hex" and RESPONSE_INIT_HEX. Removed active attack
		protection support.
	Modified by cmetz for OPIE 2.3. OPIE_PASS_MAX changed to
		OPIE_SECRET_MAX. Added extended responses, which created
		lots of changes. Eliminated extra variable. Added -x and
		-t to help. Added -f flag. Added SHA support.
	Modified by cmetz for OPIE 2.22. Print newline after seed too long
	        message. Check for minimum seed length. Correct a grammar
		error.
        Modified at NRL for OPIE 2.2. Check opiereadpass() return.
                Change opiereadpass() calls to add echo arg. Use FUNCTION
                definition et al. Check seed length here, too. Added back
		hex output. Reworked final output function.
	Modified at NRL for OPIE 2.0.
	Written at Bellcore for the S/Key Version 1 software distribution
		(skey.c).

$FreeBSD: release/9.0.0/contrib/opie/opiekey.c 92914 2002-03-21 23:42:52Z markm $

*/
#include "opie_cfg.h"

#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#include "opie.h"

#ifdef	__MSDOS__
#include <dos.h>
#endif

#if HAVE_FCNTL_H
#include <fcntl.h>
#endif /* HAVE_FCNTL_H */

extern char *optarg;
extern int optind, opterr;

int aflag = 0;

char *algnames[] = { NULL, NULL, NULL, "SHA-1", "MD4", "MD5" };
char *algids[] = { NULL, NULL, NULL, "sha1", "md4", "md5" };

/******** Begin real source code ***************/

static VOIDRET usage FUNCTION((s), char *s)
{
  fprintf(stderr, "usage: %s [-v] [-h] [-f] [-x] [-t type] [-4 | -5 | -s] [-a] [-n count] sequence_number seed\n", s);
  exit(1);
}

#define RESPONSE_STANDARD  0
#define RESPONSE_WORD      1
#define RESPONSE_HEX       2
#define RESPONSE_INIT_HEX  3
#define RESPONSE_INIT_WORD 4
#define RESPONSE_UNKNOWN   5

struct _rtrans {
  int type;
  char *name;
};

static struct _rtrans rtrans[] = {
  { RESPONSE_WORD, "word" },
  { RESPONSE_HEX, "hex" },
  { RESPONSE_INIT_HEX, "init-hex" },
  { RESPONSE_INIT_WORD, "init-word" },
  { RESPONSE_STANDARD, "" },
  { RESPONSE_STANDARD, "standard" },
  { RESPONSE_STANDARD, "otp" },
  { RESPONSE_UNKNOWN, NULL }
};

static void getsecret FUNCTION((secret, promptextra, retype), char *secret AND char *promptextra AND int flags)
{
  fprintf(stderr, "Enter %ssecret pass phrase: ", promptextra);
  if (!opiereadpass(secret, OPIE_SECRET_MAX, 0)) {
    fprintf(stderr, "Error reading %ssecret pass phrase!\n", promptextra);
    exit(1);
  }
  if (secret[0] && (flags & 1)) {
    char verify[OPIE_SECRET_MAX + 1];

    fprintf(stderr, "Again %ssecret pass phrase: ", promptextra);
    if (!opiereadpass(verify, OPIE_SECRET_MAX, 0)) {
      fprintf(stderr, "Error reading %ssecret pass phrase!\n", promptextra);
      memset(verify, 0, sizeof(verify));
      memset(secret, 0, sizeof(secret));
      exit(1);
    }
    if (verify[0] && strcmp(verify, secret)) {
      fprintf(stderr, "They don't match. Try again.\n");
      memset(verify, 0, sizeof(verify));
      memset(secret, 0, sizeof(secret));
      exit(1);
    }
    memset(verify, 0, sizeof(verify));
  }
  if (!(flags & 2) && !aflag && opiepasscheck(secret)) {
    memset(secret, 0, sizeof(secret));
    fprintf(stderr, "Secret pass phrases must be between %d and %d characters long.\n", OPIE_SECRET_MIN, OPIE_SECRET_MAX);
    exit(1);
  };
}

int main FUNCTION((argc, argv), int argc AND char *argv[])
{
  /* variable declarations */
  unsigned algorithm = MDX;	/* default algorithm per Makefile's MDX
				   symbol */
  int keynum = 0;
  int i;
  int count = 1;
  char secret[OPIE_SECRET_MAX + 1], newsecret[OPIE_SECRET_MAX + 1];
  struct opie_otpkey key, newkey;
  char *seed, newseed[OPIE_SEED_MAX + 1];
  char response[OPIE_RESPONSE_MAX + 1];
  char *slash;
  int hex = 0;
  int type = RESPONSE_STANDARD;
  int force = 0;

  if (slash = strchr(argv[0], '/'))
    slash++;
  else
    slash = argv[0];

  if (!strcmp(slash, "key") || strstr(slash, "md4"))
    algorithm = 4;

  if (strstr(slash, "md5"))
    algorithm = 5;

  if (strstr(slash, "sha"))
    algorithm = 3;

  while ((i = getopt(argc, argv, "fhvn:x45at:s")) != EOF) {
    switch (i) {
    case 'v':
      opieversion();

    case 'n':
      count = atoi(optarg);
      break;

    case 'x':
      hex = 1;
      break;

    case 'f':
#if INSECURE_OVERRIDE
      force = 1;
#else /* INSECURE_OVERRIDE */
      fprintf(stderr, "Sorry, but the -f option is not supported by this build of OPIE.\n");
#endif /* INSECURE_OVERRIDE */
      break;

    case '4':
      /* use MD4 algorithm */
      algorithm = 4;
      break;

    case '5':
      /* use MD5 algorithm */
      algorithm = 5;
      break;

    case 'a':
      aflag = 1;
      break;

    case 't':
      {
	struct _rtrans *r;
	for (r = rtrans; r->name && strcmp(r->name, optarg); r++);
	if (!r->name) {
	  fprintf(stderr, "%s: %s: unknown response type.\n", argv[0], optarg);
	  exit(1);
	}
	type = r->type;
      }
      break;

    case 's':
      algorithm = 3;
      break;

    default:
      usage(argv[0]);
    }
  }

  if ((argc - optind) < 2)
    usage(argv[0]);

  fprintf(stderr, "Using the %s algorithm to compute response.\n", algnames[algorithm]);

  /* get sequence number, which is next-to-last parameter */
  keynum = atoi(argv[optind]);
  if (keynum < 1) {
    fprintf(stderr, "Sequence number %s is not positive.\n", argv[optind]);
    exit(1);
  }
  /* get seed string, which is last parameter */
  seed = argv[optind + 1];
  {
    i = strlen(seed);

    if (i > OPIE_SEED_MAX) {
      fprintf(stderr, "Seeds must be less than %d characters long.\n", OPIE_SEED_MAX);
      exit(1);
    }
    if (i < OPIE_SEED_MIN) {
      fprintf(stderr, "Seeds must be greater than %d characters long.\n", OPIE_SEED_MIN);
      exit(1);
    }
  }

  fprintf(stderr, "Reminder: Don't use opiekey from telnet or dial-in sessions.\n");

  if (opieinsecure()) {
    fprintf(stderr, "Sorry, but you don't seem to be on the console or a secure terminal.\n");
#if INSECURE_OVERRIDE
    if (force)
      fprintf(stderr, "Warning: Continuing could disclose your secret pass phrase to an attacker!\n");
    else
#endif /* INSECURE_OVERRIDE */
      exit(1);
  }

  if ((type == RESPONSE_INIT_HEX) || (type == RESPONSE_INIT_WORD)) {
#if RETYPE
    getsecret(secret, "old ", 1);
#else /* RETYPE */
    getsecret(secret, "old ", 0);
#endif /* RETYPE */
    getsecret(newsecret, "new ", 1);
    if (!newsecret[0])
      strcpy(newsecret, secret);

    if (opienewseed(strcpy(newseed, seed)) < 0) {
      fprintf(stderr, "Error updating seed.\n");
      goto error;
    }

    if (opiekeycrunch(algorithm, &newkey, newseed, newsecret)) {
      fprintf(stderr, "%s: key crunch failed (1)\n", argv[0]);
      goto error;
    }

    for (i = 0; i < 499; i++)
      opiehash(&newkey, algorithm);
  } else
#if RETYPE
    getsecret(secret, "", 1);
#else /* RETYPE */
    getsecret(secret, "", 0);
#endif /* RETYPE */

  /* Crunch seed and secret password into starting key normally */
  if (opiekeycrunch(algorithm, &key, seed, secret)) {
    fprintf(stderr, "%s: key crunch failed\n", argv[0]);
    goto error;
  }

  for (i = 0; i <= (keynum - count); i++)
    opiehash(&key, algorithm);

  {
    char buf[OPIE_SEED_MAX + 48 + 1];
    char *c;

    for (; i <= keynum; i++) {
      if (count > 1)
	printf("%d: %s", i, (type == RESPONSE_STANDARD) ? "" : "\n");
      
      switch(type) {
      case RESPONSE_STANDARD:
	if (hex)
	  opiebtoh(response, &key);
	else
	  opiebtoe(response, &key);
	break;
      case RESPONSE_WORD:
	strcpy(response, "word:");
	strcat(response, opiebtoe(buf, &key));
	break;
      case RESPONSE_HEX:
	strcpy(response, "hex:");
	strcat(response, opiebtoh(buf, &key));
	break;
      case RESPONSE_INIT_HEX:
      case RESPONSE_INIT_WORD:
	if (type == RESPONSE_INIT_HEX) {
	  strcpy(response, "init-hex:");
	  strcat(response, opiebtoh(buf, &key));
	  sprintf(buf, ":%s 499 %s:", algids[algorithm], newseed);
	  strcat(response, buf);
	  strcat(response, opiebtoh(buf, &newkey));
	} else {
	  strcpy(response, "init-word:");
	  strcat(response, opiebtoe(buf, &key));
	  sprintf(buf, ":%s 499 %s:", algids[algorithm], newseed);
	  strcat(response, buf);
	  strcat(response, opiebtoe(buf, &newkey));
	}
	break;
      }
      puts(response);
      opiehash(&key, algorithm);
    }
  }

  memset(secret, 0, sizeof(secret));
  memset(newsecret, 0, sizeof(newsecret));
  return 0;

error:
  memset(secret, 0, sizeof(secret));
  memset(newsecret, 0, sizeof(newsecret));
  return 1;
}
